### ISSUE TITLE
UDP support

### ISSUE THREAD

**Issue by krismcqueen** [k3po/k3po#106]
UDP support in K3PO is desired.  An example of where this would be used is [specification.socks5](https://github.com/k3po/specification.socks5/issues/6) in order to implement the UDP Associate command present in the SOCKS5 protocol.  This would mean K3PO would support binding to a UDP address and receiving datagram packets.

**Comment by dpwspoon** [comment]
Mina.netty has a [Connectionless bootstrap](https://netty.io/3.9/api/org/jboss/netty/bootstrap/ConnectionlessBootstrap.html) for UDP.  We have always treated k3po scripts as server or client with corresponding `accept` or `connect`.  In UDP does it make sense to correlate UDP packets in a script as 

```
accept udp://   or connect udp://

close
closed
```

Or would a different non-client/server paradigm be better?

**Comment by tmontgomery** [comment]
A `connect()` for UDP sockets means associate a given IP address and port with this socket. So, all data sent to that socket that doesn't come from the IP/port is dropped. And `send()` always goes to that IP/port.

More semantics... `sendto()` includes the IP/port each time it is called. And `recvfrom()` returns the IP/port of the sender.

I can see use cases where the equivalent of `sendto()` and `recvfrom()` are very handy with K3PO.

**Comment by dpwspoon** [comment]
If I understand correctly, then it sounds like we could have `connect udp://local-address udp://remote-address` which would ignore all data not sent from the remote-address.  Maybe there is a better formatting for the address layout?

And `accept udp://listen-address` where something like `read "Some data" from udp://remote-address` and `write "Some data" to udp://remote-address`  would be the equivalent to sendto() and recvfrom(). `to and `read` would be new keywords in the language.  

Would it make sense then to have reads and writes in an `accept: upd://  ` be syntactically invalid?

**Comment by rlankenau** [comment]
I feel like connect() semantics in a UDP socket is a bit limiting.  I think I'd expect to simply open a socket on a given IP/port, and then specify expected reads as a tuple including data, source IP and source port.

On the send side, specifying a destination IP, destination port and payload doesn't seem that much more difficult than specifying a connect way up and the top.

**Comment by tmontgomery** [comment]
Could the `to` and `from` use variable substitution so that you can assign IP/port values to variables like `"receiver A"`, `"topic sender"`, etc.

**Comment by rlankenau** [comment]
Something else to think about is that you might want to be able to mark the end of a datagram to check that the datagram contains exactly what you expect and no more.

@tmontgomery If we add first class Aeron support to K3P0 some time in the future (something we've talked about here in Warrenville), I would think that we'd want something similar to the UDP interface so you could check completeness of messages.

**Comment by dpwspoon** [comment]
@tmontgomery Good point! The `to` and `from` should support properties/variables or addresses.  Also for readability perhaps they should probably be directly after the read/write

@rlankenau, so it sounds like perhaps all that is needed is an `open` keyword, and we could ditch `accept` and `connect`.  Sample UDP Script

```

property endpoint1 udp://127.0.0.1:5051
property endpoint2 udp://127.0.0.1:5052

# start endpoint #1 channel via property
open endpoint1
write to endpoint2 "Hello"
read from endpoint2 "World"

# write and read to another address 
read from udp://127.0.0.1:5050 "Some Data text data followed by bytes" [0x12 0x3F]
write to udp://127.0.0.1:5050 "Some response to data"

# close is sufficient to signal the end of udp stream
close

# start endpoint #2 (could use property but for example use address)
open udp://127.0.0.1:5052
read to endpoint1 "Hello"
write from endpoint1 "World"
close

```

**Comment by dpwspoon** [comment]
I guess we still need to distinguish between bind/Listening.  Perhaps the semantics should more mimic the [datagram API](http://download.java.net/jdk7/archive/b123/docs/api/java/net/DatagramSocket.html)

**Comment by tmontgomery** [comment]
`close` semantics for UDP might be a bit odd. There is no equivalent EOF indication.

**Comment by tmontgomery** [comment]
@rlankenau I think the semantics of `read` for K3PO will do the exact match you mention. But @dpwspoon may correct me on that.

**Comment by tmontgomery** [comment]
@dpwspoon correct, the receiving side will need `bind` semantics.

For multicast reception, it will need specific `IP_ADD_MEMBERSHIP` semantics. Which includes interface, group, and destination port.

For multicast transmission, it should be enough to specify a group, interface, and destination port.

Both of the above will need the ability to specify interface separately from the group address.

**Comment by dpwspoon** [comment]
I suggest using the same semantics as we already have in the gateway and pairing connections on the sourceIP:sourcePort, destinationIp:destinationPort tuple.

i.e. 

```
connect udp://accept-address:accept-port
connected

closed

accept udp://connect-address:connect-port
accepted 
connected

close
closed

```
a `closed` would be noted after a set amount of time has passed with no data. 
That amount of time would default to say 5 seconds.  And be configurable via an accept/connect option.

a `close` would be the equivilant of not listening anymore
